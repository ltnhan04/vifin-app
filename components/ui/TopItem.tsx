import { View, Text } from "react-native";

const TopItem = ({
  title,
  price,
  percent,
  type,
}: {
  title: string;
  price: number;
  percent: number;
  type: "expense" | "income";
}) => {
  return (
    <View className="mt-2">
      <Text className="text-base text-white font-rubik-medium">{title}</Text>
      <View className="flex flex-row items-center justify-between">
        <Text className="text-sm text-white font-rubik-medium">
          {price} VND
        </Text>
        <Text
          className={`${type === "expense" ? "text-secondary-red" : "text-secondary-green-100"} font-rubik-bold text-sm`}
        >
          {percent}%
        </Text>
      </View>
    </View>
  );
};

export default TopItem;
