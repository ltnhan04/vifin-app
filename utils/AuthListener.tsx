import { useState, useEffect } from "react";
import { router } from "expo-router";
import auth, { FirebaseAuthTypes } from "@react-native-firebase/auth";
import { useAppDispatch } from "@/redux/hooks";
import { setUser, clearUser } from "@/redux/features/auth/authSlice";

const AuthListener = ({ children }: { children: React.ReactNode }) => {
  const dispatch = useAppDispatch();
  const [initializing, setInitializing] = useState(true);

  const onAuthStateChanged = async (user: FirebaseAuthTypes.User | null) => {
    if (user) {
      const tokenResult = await user.getIdTokenResult();
      const expirationTime = new Date(tokenResult.expirationTime);
      if (expirationTime < new Date()) {
        dispatch(clearUser());
        router.push("/(auth)/sign-in");
      } else {
        dispatch(setUser({ user: user, token: tokenResult.token }));
        router.replace("/(tabs)/home");
      }
    } else {
      dispatch(clearUser());
      router.replace("/(auth)/sign-in");
    }
    if (initializing) setInitializing(false);
  };
  useEffect(() => {
    const subscriber = auth().onAuthStateChanged(onAuthStateChanged);
    return subscriber;
  }, []);
  return children;
};

export default AuthListener;
