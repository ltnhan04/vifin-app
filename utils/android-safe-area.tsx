import { StyleSheet, Platform } from "react-native";

export default StyleSheet.create({
  androidSafeArea: {
    flex: 1,
    paddingTop: Platform.OS === "android" ? 24 : 0,
  },
});
