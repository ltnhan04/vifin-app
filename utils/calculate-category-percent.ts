import { categoryColors } from "@/constants/data";
import { ITransaction } from "@/types/transaction";
interface IPieData {
  value: number;
  color: string;
  text: string;
  icon: string;
}

const calculatePercent = (total: number, amount: number) => {
  return Math.round((amount / total) * 100);
};

export const CalculateCategoryPercent = (
  total: number,
  transactions: ITransaction[]
) => {
  const pieData: IPieData[] = [];

  transactions.map((transaction) => {
    const categoryData = transaction.category;
    const value = calculatePercent(total, transaction.amount);
    const findColor = categoryColors.find(
      (category) => category.category === categoryData.name
    );
    pieData.push({
      value,
      color: findColor?.color ?? "#CCCCCC",
      text: `${value}%`,
      icon: categoryData.symbol,
    });
  });

  return pieData;
};
