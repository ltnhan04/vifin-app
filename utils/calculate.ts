import { IBudgetByRepeatType } from "@/types/budget";

export const calculateBudget = (budgetData: IBudgetByRepeatType[]) => {
  const totalLimit = budgetData.reduce(
    (sum, budget) => sum + (budget.amount || 0),
    0
  );
  const totalSpent = budgetData.reduce(
    (sum, budget) => sum + (budget.usage || 0),
    0
  );
  const percentage = totalLimit > 0 ? (totalSpent / totalLimit) * 100 : 0;

  return { totalLimit, totalSpent, percentage };
};

export const calculateExpense = (goalAmount: number, currentAmount: number) => {
  const remainingAmount = goalAmount - currentAmount;
  const progress = currentAmount / goalAmount;
  const percentage = Math.round(progress * 100);
  return { remainingAmount, percentage, progress };
};
