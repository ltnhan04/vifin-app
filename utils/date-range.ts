const now = new Date();
const weekStart = new Date(now);
const weekEnd = new Date(now);
weekEnd.setDate(now.getDate() + 7);

const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
const monthEnd = new Date(now.getFullYear(), now.getMonth() + 1, 0);

const yearStart = new Date(now.getFullYear(), 0, 1);
const yearEnd = new Date(now.getFullYear(), 11, 31);
export { weekStart, weekEnd, monthStart, monthEnd, yearStart, yearEnd };
