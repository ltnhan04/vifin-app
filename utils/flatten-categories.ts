export const flattenCategories = (categories: any[], prefix = "") => {
  let flatList: any[] = [];
  categories.forEach((category) => {
    flatList.push({
      _id: category._id,
      name: `${prefix}${category.name}`,
    });
    if (category.children && category.children.length > 0) {
      flatList = flatList.concat(
        flattenCategories(category.children, prefix + "--- ")
      );
    }
  });
  return flatList;
};
