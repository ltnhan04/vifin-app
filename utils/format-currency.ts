type Currency = "VND" | "USD";
export const formatCurrency = (amount: number, currency: Currency) => {
  let locate;
  switch (currency) {
    case "VND":
      locate = "vi-VN";
      break;
    case "USD":
      locate = "en-US";
      break;
    default:
      locate = "vi-VN";
  }

  return new Intl.NumberFormat(locate, {
    style: "currency",
    currency: currency,
    minimumFractionDigits: currency === "VND" ? 0 : 2,
  }).format(amount);
};

export const formatValue = (value: number): string => {
  if (value >= 1000000) {
    return `${(value / 1000000).toFixed(1)}M`;
  } else if (value >= 1000) {
    return `${(value / 1000).toFixed(1)}K`;
  }
  return value.toString();
};
