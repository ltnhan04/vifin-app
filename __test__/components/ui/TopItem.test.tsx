import React from "react";
import { render } from "@testing-library/react-native";
import { Image } from "react-native";
import TopItem from "@/components/ui/TopItem";
import { formatCurrency } from "@/utils/format-currency";
import type { IconKey } from "@/types/icon";

// Mock hàm formatCurrency để kiểm soát output định dạng
jest.mock("@/utils/format-currency", () => ({
  formatCurrency: jest.fn((price, currency) => `${price} ${currency}`),
}));

// Mock module icons để trả về một source đơn giản cho icon
jest.mock("@/constants/icons", () => ({
  walletIcon: { uri: "dummy-wallet-icon" },
}));

describe("TopItem Component", () => {
  const defaultProps = {
    title: "Test Title",
    price: 1000,
    percent: 10,
    type: "expense" as "expense" | "income",
    avt: "walletIcon" as IconKey, // ép kiểu giá trị "walletIcon" thành IconKey
  };

  it("renders title, formatted price and percent correctly", () => {
    const { getByText } = render(<TopItem {...defaultProps} />);
    expect(getByText("Test Title")).toBeTruthy();
    // Vì hàm formatCurrency được mock trả về "1000 VND"
    expect(getByText("1000 VND")).toBeTruthy();
    expect(getByText("10%")).toBeTruthy();
  });

  it("renders icon image with correct source", () => {
    const { UNSAFE_getByType } = render(<TopItem {...defaultProps} />);
    const image = UNSAFE_getByType(Image);
    expect(image.props.source).toEqual({ uri: "dummy-wallet-icon" });
  });

  it("renders correct percent style based on type", () => {
    // Sử dụng snapshot testing để so sánh giao diện khi type là expense và income
    const { toJSON, rerender } = render(<TopItem {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();

    rerender(<TopItem {...defaultProps} type="income" />);
    expect(toJSON()).toMatchSnapshot();
  });
});
