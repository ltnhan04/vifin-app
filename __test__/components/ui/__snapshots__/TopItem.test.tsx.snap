// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopItem Component renders correct percent style based on type 1`] = `
<View
  className="flex flex-row items-center gap-x-3"
>
  <Image
    className="size-9 bg-primary-dark rounded-full"
    source={
      {
        "uri": "dummy-wallet-icon",
      }
    }
  />
  <View
    className="mt-2 flex-1"
  >
    <Text
      className="text-base text-white font-rubik-medium"
    >
      Test Title
    </Text>
    <View
      className="flex flex-row items-center justify-between"
    >
      <Text
        className="text-sm text-white font-rubik-medium"
      >
        1000 VND
      </Text>
      <Text
        className="text-secondary-red font-rubik-bold text-sm"
      >
        10
        %
      </Text>
    </View>
  </View>
</View>
`;

exports[`TopItem Component renders correct percent style based on type 2`] = `
<View
  className="flex flex-row items-center gap-x-3"
>
  <Image
    className="size-9 bg-primary-dark rounded-full"
    source={
      {
        "uri": "dummy-wallet-icon",
      }
    }
  />
  <View
    className="mt-2 flex-1"
  >
    <Text
      className="text-base text-white font-rubik-medium"
    >
      Test Title
    </Text>
    <View
      className="flex flex-row items-center justify-between"
    >
      <Text
        className="text-sm text-white font-rubik-medium"
      >
        1000 VND
      </Text>
      <Text
        className="text-secondary-green-100 font-rubik-bold text-sm"
      >
        10
        %
      </Text>
    </View>
  </View>
</View>
`;
